### PR TITLE
Improving CacheGuardian to free evicted fibers as soon as possible

### PR DESCRIPTION
**Pull request by lee-lei**
## What changes were proposed in this pull request?

This PR is improving the cache guardian thread in order to:
1. ensure the pending fibers in the evicted queue to be freed as soon as possible
   after it's released by the last user.
2. ensure the queue to be empty as long as all the pending fibers are released by the last users.
With the above assurances, the memory pressure will be significantly mitigated.

We are based on the premise that the guava cache manager has a good eviction mechanism.
That indicates that the evicted fibers are relatively old and not accessed for a while and
won't be accessed in the future with the high possibility. Otherwise, we will upload another PR to address the inefficient eviction algorithm.

## How was this patch tested?

mvn clean test package and all the unit tests passed.

For the total 13 existing fiber cache manager test cases, the peak memory footprint of five are reduced significantly. The maximum memory size of cache manager is 7.34MB.

The detailed result is below:
The peak memory of "remove a fiber is in use" is reduced from 66.06MB to under 7.34MB.
The peak memory of "wait for other thread to release the fiber" is reduced from 31.46MB to under 7.34MB.
The peak memory of "add a very large fiber" is reduced from 39.32MB to under 7.34MB.
The peak memory of "cache guardian remove pending fibers" is reduced from 76.55MB to 69.21MB.
The peak memory of "get different fiber simultaneously" is reduced from 33.55MB to 12.58MB.

### CODE REVIEW

**Comment by lee-lei** [comment]
For this case, the guava cache manager will directly free the evicted fiber if the refCount is zero already before putting in the queue. The code is in addRemovalFiber method in the first patch. Thanks.

**Comment by lee-lei** [comment]
@carsonwang Please help review if the third commit resolves your concern when you are available. Thanks so much for your comments!

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCache.scala [hunk 1]
```diff
@@ -44,49 +51,19 @@ case class FiberCache(protected val fiberData: MemoryBlock) extends Logging {
...
     }
-    logWarning(s"Fiber Cache Dispose waiting detected for $fiber")
-    false
   }
 
-  protected var disposed = false
-  def isDisposed: Boolean = disposed
-  protected[filecache] def realDispose(fiber: Fiber): Unit = {
+  private var disposed: Boolean = false
+  def isDisposed(): Boolean = disposed
```
Comment (carsonwang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCache.scala]: It is ok without () here because there is no side effect for this method.

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCache.scala [hunk 2]
```diff
@@ -28,7 +29,7 @@ import org.apache.spark.unsafe.memory.MemoryBlock
 import org.apache.spark.unsafe.types.UTF8String
 
 // TODO: make it an alias of MemoryBlock
-case class FiberCache(protected val fiberData: MemoryBlock) extends Logging {
+case class FiberCache(var fiber: Fiber, protected val fiberData: MemoryBlock) extends Logging {
```
Comment (gczsjdy) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCache.scala]: For the changes of main interfaces/classes, please indicate your design and the necessity

**Comment by gczsjdy** [comment]
You probably should indicate the direct evidence of why current cache guardian thread doesn't release fiber cache well(and as you say, cause memory pressure)

For your test part, some cases of `FiberCacheManagerSuite`'s producing `fibers pending on removal use too much memory` can't prove the inefficiency of `CacheGuardian`. So you can't prove your PR works by 

>   the test cases exceeding off-heap memory usage threshold are reduced from 6 to 2

Because they are made to overuse the memory in a sudden to let Guava perform evict, so it's likely the pending queue is temporarily full. And all of that happen in a rather short time.

In short, please
- Prove the evidence of memory pressure, maybe real cases and statistics (and it's caused by CacheGuardian)
- Use PR on the cases indicated above and give new results

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCache.scala (hunk 2 above)
Comment (lee-lei) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCache.scala]: For the last commit, there are primarily two changes. 
The first change is to add the fiber information to the corresponding fiber cache. The logic is that
we want to let the last user of fiber cache to release the fiber and free the memory together. Thus the fiber cache needs the information that it belongs to which fiber.
The second change is to refactor the test cases to use one unified apply function of FiberCache object. In this case, it's more clean.

**Comment by lee-lei** [comment]
@gczsjdy For the unit test result, I updated in this PR. Thanks.

**Comment by gczsjdy** [comment]
@lee-lei That's in extreme conditions of UT we created(like, probably occurs in 0.01% cases), however you still hasn't proved the memory pressure in non-extreme cases.

**Comment by gczsjdy** [comment]
A complete logic chain proving the memory pressure + the result after resolving are needed

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCache.scala (hunk 2 above)
Comment (carsonwang) [reply src/main/scala/org/apache/spark/sql/execution/datasources/oap/filecache/FiberCache.scala]: `fiber` should be a val ideally. Any plan to update this?

**Comment by lee-lei** [comment]
@carsonwang please help review the last two commits when you are available. I am totally ok to split the last commit as a standalone PR after your review. Thanks.

**Comment by gczsjdy** [comment]
@lee-lei maybe it's better to split this first, for the convenience of reviewers.
Also, we need confirmation from @linhong-intel

**Comment by linhong-intel** [comment]
@gczsjdy this pr is ok to me

**Comment by lee-lei** [comment]
Close this PR, since the problem resolved by this PR is very rare in the production environment, and the #830 proposed another solution. Thanks.
